In [33]:
import numpy as np
from sklearn import decomposition
from sklearn import linear_model
import timeit

In [25]:
train = np.genfromtxt('data/gisette/gisette_train.data', delimiter=' ')
train_labels = np.genfromtxt('data/gisette/gisette_train.labels')

valid = np.genfromtxt('data/gisette/gisette_valid.data', delimiter=' ')
valid_labels = np.genfromtxt('data/gisette/gisette_valid.labels')

In [27]:
start = timeit.default_timer()

logreg = linear_model.LogisticRegression(C=1e5)
logreg.fit(train, train_labels)

stop = timeit.default_timer()

print('Time: ', stop - start)  

Time:  1.835482795999269


In [31]:
logreg.score(valid, valid_labels)

0.983

In [46]:
pca = decomposition.PCA(n_components=5)
train_c5 = pca.fit_transform(train)
valid_c5 = pca.fit_transform(valid)

pca = decomposition.PCA(n_components=10)
train_c10 = pca.fit_transform(train)
valid_c10 = pca.fit_transform(valid)

pca = decomposition.PCA(n_components=20)
train_c20 = pca.fit_transform(train)
valid_c20 = pca.fit_transform(valid)

pca = decomposition.PCA(n_components=50)
train_c50 = pca.fit_transform(train)
valid_c50 = pca.fit_transform(valid)

pca = decomposition.PCA(n_components=100)
train_c100 = pca.fit_transform(train)
valid_c100 = pca.fit_transform(valid)

In [47]:
train_c5

array([[10625.49032311,  2640.71752336,  6733.28222802,  8270.69418531,
         1052.48136627],
       [ 5046.23858128, -5301.98908288, -4247.26914654, -2492.18409595,
          194.59734199],
       [ 1814.15443383, -5894.79062872, -2199.47862341,  2353.53518526,
         3805.57015389],
       ...,
       [  875.08216412, -5584.87438239, -2768.44674207, -1023.47457228,
         -512.865912  ],
       [-1065.71292402, -4327.67189742,  2537.08913045,  -938.08085425,
         -330.82203484],
       [-1307.07497986,  5726.6343569 , -3408.67005296, -2835.41206935,
         3140.72346736]])

In [58]:
start = timeit.default_timer()

logreg_c5 = linear_model.LogisticRegression(C=1e5)
logreg_c5.fit(train_c5, train_labels)

stop = timeit.default_timer()

print('Time for n=5: ', stop - start)  
print('Score for n=5: ', logreg_c5.score(valid_c5, valid_labels))

Time for n=5:  0.0076652810093946755
Score for n=5:  0.785


In [59]:
start = timeit.default_timer()

logreg_c10 = linear_model.LogisticRegression(C=1e5)
logreg_c10.fit(train_c10, train_labels)

stop = timeit.default_timer()

print('Time for n=10: ', stop - start)  
print('Score for n=10: ', logreg_c10.score(valid_c10, valid_labels))

Time for n=10:  0.011238135994062759
Score for n=10:  0.77


In [60]:
start = timeit.default_timer()

logreg_c20 = linear_model.LogisticRegression(C=1e5)
logreg_c20.fit(train_c20, train_labels)

stop = timeit.default_timer()

print('Time for n=20: ', stop - start)  
print('Score for n=20: ', logreg_c20.score(valid_c20, valid_labels))

Time for n=20:  0.026845837011933327
Score for n=20:  0.774


In [61]:
start = timeit.default_timer()

logreg_c50 = linear_model.LogisticRegression(C=1e5)
logreg_c50.fit(train_c50, train_labels)

stop = timeit.default_timer()

print('Time for n=50: ', stop - start)  
print('Score for n=50: ', logreg_c50.score(valid_c50, valid_labels))

Time for n=50:  0.06504962699546013
Score for n=50:  0.762


In [62]:
start = timeit.default_timer()

logreg_c100 = linear_model.LogisticRegression(C=1e5)
logreg_c100.fit(train_c100, train_labels)

stop = timeit.default_timer()

print('Time for n=100: ', stop - start)  
print('Score for n=100: ', logreg_c100.score(valid_c100, valid_labels))

Time for n=100:  0.18855567500577308
Score for n=100:  0.758


In [66]:
transformer = decomposition.FactorAnalysis(n_components=3, random_state=0)
train_c3f = transformer.fit_transform(train)
valid_c3f = transformer.fit_transform(valid)

transformer = decomposition.FactorAnalysis(n_components=5, random_state=0)
train_c5f = transformer.fit_transform(train)
valid_c5f = transformer.fit_transform(valid)

transformer = decomposition.FactorAnalysis(n_components=10, random_state=0)
train_c10f = transformer.fit_transform(train)
valid_c10f = transformer.fit_transform(valid)

transformer = decomposition.FactorAnalysis(n_components=20, random_state=0)
train_c20f = transformer.fit_transform(train)
valid_c20f = transformer.fit_transform(valid)

transformer = decomposition.FactorAnalysis(n_components=50, random_state=0)
train_c50f = transformer.fit_transform(train)
valid_c50f = transformer.fit_transform(valid)

In [68]:
start = timeit.default_timer()

logreg_c3f = linear_model.LogisticRegression(C=1e5)
logreg_c3f.fit(train_c3f, train_labels)

stop = timeit.default_timer()

print('Time for n=3: ', stop - start)  
print('Score for n=3: ', logreg_c3f.score(valid_c3f, valid_labels))

start = timeit.default_timer()

logreg_c5f = linear_model.LogisticRegression(C=1e5)
logreg_c5f.fit(train_c5f, train_labels)

stop = timeit.default_timer()

print('Time for n=5: ', stop - start)  
print('Score for n=5: ', logreg_c5f.score(valid_c5f, valid_labels))

start = timeit.default_timer()

logreg_c10f = linear_model.LogisticRegression(C=1e5)
logreg_c10f.fit(train_c10f, train_labels)

stop = timeit.default_timer()

print('Time for n=10: ', stop - start)  
print('Score for n=10: ', logreg_c10f.score(valid_c10f, valid_labels))

start = timeit.default_timer()

logreg_c20f = linear_model.LogisticRegression(C=1e5)
logreg_c20f.fit(train_c20f, train_labels)

stop = timeit.default_timer()

print('Time for n=20: ', stop - start)  
print('Score for n=20: ', logreg_c20f.score(valid_c20f, valid_labels))

start = timeit.default_timer()

logreg_c50f = linear_model.LogisticRegression(C=1e5)
logreg_c50f.fit(train_c50f, train_labels)

stop = timeit.default_timer()

print('Time for n=50: ', stop - start)  
print('Score for n=50: ', logreg_c50f.score(valid_c50f, valid_labels))

Time for n=3:  0.007386415003566071
Score for n=3:  0.744
Time for n=5:  0.006738064999808557
Score for n=5:  0.812
Time for n=10:  0.008673792995978147
Score for n=10:  0.856
Time for n=20:  0.018179655002313666
Score for n=20:  0.76
Time for n=50:  0.03418337400944438
Score for n=50:  0.695


## Original model

Time:  1.835482795999269
Score: 0.983

Pros:  
good accuracy  
Cons:  
long training  

## Best score from PCA

Time for n=5:  0.0076652810093946755
Score for n=5:  0.785

Pros:  
very efficient  
Cons:  
Strongly focused in variance

## Best score from FA

Time for n=10:  0.008673792995978147
Score for n=10:  0.856
